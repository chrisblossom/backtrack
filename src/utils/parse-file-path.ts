import path from 'path';
import { existsSync } from 'fs';
import { getFileHash } from './get-file-hash';

import { rootPath } from '../config/paths';

type Return = Readonly<{
    absolute: string;
    relative: string;
    hash: string;
}>;

function parseFilePath(file: string): Return {
    const absolute = path.resolve(rootPath, file);
    const relative = path.relative(rootPath, absolute);

    const exists = existsSync(absolute);
    if (exists === false) {
        return {
            absolute,
            relative,
            hash: '',
        };
    }

    const hash = getFileHash(absolute);
    return {
        absolute,
        relative,
        hash,
    };
}

export { parseFilePath };
