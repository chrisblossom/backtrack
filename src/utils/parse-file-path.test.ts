import path from 'path';

require('./parse-file-path');

const parseFilePath = (file: string) =>
    require('./parse-file-path').parseFilePath(file);

describe('parseFilePath', () => {
    const cwd = process.cwd();
    let dir: string;

    beforeEach(() => {
        dir = path.resolve(__dirname, '__sandbox__/files1/');
        process.chdir(dir);
    });

    afterEach(() => {
        process.chdir(cwd);
    });

    test('parses absolute file', () => {
        const file = path.resolve(dir, 'test-file.js');

        const result = parseFilePath(file);

        expect(result).toMatchSnapshot();
    });

    test('parses relative file', () => {
        const file = 'test-file.js';

        const result = parseFilePath(file);

        expect(result).toMatchSnapshot();
    });

    test('parses nested relative file', () => {
        const file = 'nested/other.js';

        const result = parseFilePath(file);

        expect(result).toMatchSnapshot();
    });

    test('parses nested absolute file', () => {
        const file = path.resolve(dir, 'nested/other.js');

        const result = parseFilePath(file);

        expect(result).toMatchSnapshot();
    });

    test('parses ./ relative file', () => {
        const file = path.resolve(dir, './nested/other.js');

        const result = parseFilePath(file);

        expect(result).toMatchSnapshot();
    });

    test('returns empty string when file does not exist', () => {
        const file = path.resolve(dir, './missing-file.js');

        const result = parseFilePath(file);

        expect(result).toMatchSnapshot();
    });
});
