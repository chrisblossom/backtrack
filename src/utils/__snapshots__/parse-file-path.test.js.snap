// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFilePath parses ./ relative file 1`] = `
Object {
  "absolute": "<PROJECT_ROOT>/nested/other.js",
  "hash": "4a98043a27f1a81a69af211207912f39873a6fe693be4cedb6bebd33cd3f2da7",
  "relative": "nested/other.js",
}
`;

exports[`parseFilePath parses absolute file 1`] = `
Object {
  "absolute": "<PROJECT_ROOT>/test-file.js",
  "hash": "b2c22cd66511e73bc0c189cc169bf13ef5ffd2fc1c988b5f0eb0cbf992456767",
  "relative": "test-file.js",
}
`;

exports[`parseFilePath parses nested absolute file 1`] = `
Object {
  "absolute": "<PROJECT_ROOT>/nested/other.js",
  "hash": "4a98043a27f1a81a69af211207912f39873a6fe693be4cedb6bebd33cd3f2da7",
  "relative": "nested/other.js",
}
`;

exports[`parseFilePath parses nested relative file 1`] = `
Object {
  "absolute": "<PROJECT_ROOT>/nested/other.js",
  "hash": "4a98043a27f1a81a69af211207912f39873a6fe693be4cedb6bebd33cd3f2da7",
  "relative": "nested/other.js",
}
`;

exports[`parseFilePath parses relative file 1`] = `
Object {
  "absolute": "<PROJECT_ROOT>/test-file.js",
  "hash": "b2c22cd66511e73bc0c189cc169bf13ef5ffd2fc1c988b5f0eb0cbf992456767",
  "relative": "test-file.js",
}
`;

exports[`parseFilePath returns empty string when file does not exist 1`] = `
Object {
  "absolute": "<PROJECT_ROOT>/missing-file.js",
  "hash": "",
  "relative": "missing-file.js",
}
`;
